File: src/java_impl/concreteTimeFormats/VerbalTimeFormat.java
package java_impl.concreteTimeFormats;

import java_impl.TimeFormat;

/**
 * Example format output:
 *
 * 1) 2 minutes and 43 seconds
 * 2) 2 hours, 52 minutes and 37 seconds
 * 3) 5 months, 32 minutes and 2 seconds
 * 4) 1 year, 2 months, 12 days, 4 hours, 29 minutes and 55 seconds
 *
 */
public class VerbalTimeFormat extends TimeFormat {

    private static VerbalTimeFormat instance;

    /**
     * @return object instance of the class
     */
    public static VerbalTimeFormat getInstance()
    {
        if(instance == null)
        {
            instance = new VerbalTimeFormat();
        }

        return instance;
    }

    /**
     * Primary method made private in order to enable the Singleton design pattern.
     */
    private VerbalTimeFormat(){}

    @Override
    public String formatSeconds(long seconds) {
        // making it pretty
        if(seconds==1)
            return "1 second";

        return seconds + " seconds";
    }

    @Override
    public String formatMinutes(long minutes) {
        // making it pretty
        if(minutes==1)
            return "1 minute";

        return minutes + " minutes";
    }

    @Override
    public String formatHours(long hours) {
        // making it pretty
        if(hours==1)
            return "1 hour";

        return hours + " hours";
    }

    @Override
    public String formatDays(long days) {
        // making it pretty
        if(days==1)
            return "1 day";

        return days + " days";
    }

    @Override
    public String formatMonths(long months) {
        // making it pretty
        if(months==1)
            return "1 month";

        return months + " months";
    }

    @Override
    public String formatYears(long years) {
        // making it pretty
        if(years==1)
            return "1 year";

        return years + " years";
    }

    @Override
    public String getJoiningCharacter() {
        return ", ";
    }

    @Override
    public String getLastJoiningCharacter()
    {
        return " and ";
    }
}

File: src/kotlin_impl/concrete_format_types/VerbalTimeFormat.kt
package kotlin_impl.concrete_format_types

import kotlin_impl.TimeFormat

/**
 * Example output formats :
 *
 * 1) xxx days, xx hours, xx minutes and xx seconds
 * - 102 days, 22 hours, 33 minutes and 21 seconds
 *
 * 2) xx hours, xx minutes and xx seconds
 * - 15 hours, 21 minutes and 44 seconds
 *
 * 3) xx minutes and xx seconds
 * - 13 minutes and 55 seconds
 *
 * 4) xx seconds
 * - 32 seconds
 */
object VerbalTimeFormat : TimeFormat() {

    override fun formatSeconds(seconds: Long): String {
        // make it pretty
        return if(seconds==1L)
            "$seconds second"
        else
            return "$seconds seconds"

    }

    override fun formatMinutes(minutes: Long): String {

        return "$minutes minutes"
    }

    override fun formatHours(hours: Long): String {

        return "$hours hours"
    }

    override fun formatDays(days: Long): String {

        return "$days days"
    }

    override fun formatMonths(months: Long): String {

        return "$months months"
    }

    override fun formatYears(years: Long): String {

        return "$years years"
    }

    override fun getJoiningString(): String {
        return ", "
    }

    override fun getLastJoiningString(): String {
        return " and "
    }
}